src/system/SystemLayer.cpp

```diff
@@ -303,6 +303,49 @@ Error Layer::StartTimer(uint32_t aMilliseconds, TimerCompleteFunct aComplete, vo
     return lReturn;
 }
 
+/**
+ * @brief
+ *   This method extends a one-shot timer.
+ *
+ *   @note
+ *       Only a single timer is allowed to be started with the same @a aComplete and @a aAppState
+ *       arguments. If called with @a aComplete and @a aAppState identical to an existing timer,
+ *       the currently-running timer will be extended for a duration of specified period.
+ *
+ *   @param[in]  aMilliseconds Extension time in milliseconds.
+ *   @param[in]  aComplete     A pointer to the function called when timer expires.
+ *   @param[in]  aAppState     A pointer to the application state object used when timer expires.
+ *
+ *   @return CHIP_SYSTEM_NO_ERROR On success.
+ *   @return CHIP_SYSTEM_ERROR_NO_MEMORY If a timer cannot be allocated.
+ *   @return CHIP_SYSTEM_ERROR_UNEXPECTED_STATE If System Layer is not initialized.
+ *   @return Other Value indicating timer failed to start or extend.
+ *
+ */
+Error Layer::ExtendTimer(uint32_t aMilliseconds, Layer::TimerCompleteFunct aOnComplete, void * aAppState)
+{
+    VerifyOrReturnError(this->State() == kLayerState_Initialized, CHIP_SYSTEM_ERROR_UNEXPECTED_STATE);
+
+    Timer * lTimer = nullptr;
+    for (size_t i = 0; i < Timer::sPool.Size(); ++i)
+    {
+        Timer * timer = Timer::sPool.Get(*this, i);
+
+        if (timer != nullptr && timer->OnComplete == aOnComplete && timer->AppState == aAppState)
+        {
+            lTimer = timer;
+            break;
+        }
+    }
+
+    if (lTimer)
+    {
+        aMilliseconds += static_cast<uint32_t>(lTimer->mAwakenEpoch - Timer::GetCurrentEpoch());
```

Comment:
I'm a little worried about how this behaves if the matching timer is overdue (i.e. if `mAwakenEpoch < GetCurrentEpoch()`).  Can that situation arise?  I don't see anything obvious preventing it.

Concretely, let's say a timer is 5ms overdue and we are trying to extend it by 2ms.  We pass in 2, add `0xFFFFFFFB`, get `0xFFFFFFFD` as the new delay.   We pass this to `StartTimer`, which casts to uint64_, adds in the current epoch, and we get a delay for a very long time from now.

I _think_ this is only an issue if we are trying to extend by less than the timer is overdue by; otherwise the addition here will overflow and we will end up with a correct extension again.... but we should this case correctly, still; otherwise we might randomly get timers that never end up firing when they should.

We should be able to add a test for this if we busy-wait for several ms in the test between first starting  the timer and extending it, right?

Reply:
Ah yes, I think the issue @bzbarsky-apple raises is real.

In addition to underflow, this code also does not check for overflow.  I know that seems theoretical, but that leads me to an additional question: is it proper to ever extend a timer in this way?  Why not just always call `StartTimer` to extend the timer to a _fixed_ time in the future?

This would avoid situations where a piece of code periodically extends a timer such that it becomes scheduled much further in the future than is intended.

And why wouldn't it work to always call `StartTimer`?  The PR description calls out this usage:

> Such as, Arm Fail-safe command need to extend the existing fail-safe timer for a during of specified period.

That sounds like a watchdog-like function.  Why can't the logic be:

* operation isn't done yet
* therefore, schedule to fire 50ms from now

Why does logic like this need to depend on the current timer value?

Reply:
So in the spec, Arm Fail-safe has these semantics:

> Upon receipt, if there isn’t already an armed fail-safe timer associated with the current secure session, a new fail-safe timer with an expiry time set for a duration of ExpiryLengthSeconds SHALL be armed. Otherwise, the existing fail-safe timer SHALL be extended for a duration of ExpiryLengthSeconds.

so in that sense we need some concept of extending timers, whether it's in the timer code itself or in the "Arm Fail-safe" implementation...

Reply:
I added the overdue check before extending, we don't know how much time still left if not implement it in timer code.

Reply:
I'm a little confused, just from @bzbarsky-apple's comments. From an SDK perspective, it seems confusing that extend doesn't fail if we don't have a timer already, can we fix that?

Reply:
This design is on purpose, since there is no API to check if the current running timer is still alive. 

Instead, we need to introduce API to tell app if the current running timer is still alive and its remaining active time.

Reply:
> @woody-apple We need to add a new API to find the timer object if we want to pass timer object to ExtendTimer, this is the APIs I proposed originally.
> 
> Timer * timer FindTimer(TimerCompleteFunct aOnComplete, void * aAppState);
> Error ExtendTimer(uint32_t aMilliseconds, Timer * timer);
> 
> Andrey's suggestion is to just return a dedicated error CHIP_SYSTEM_ERROR_TIMER_NOT_FOUND if the timer does not exist. This simplify the API and also align with existing StartTimer API.

I agree with @andy31415 , they should align.

I'm referring to this:
```
As a reader of an API of 'ExtendTimer' I would expect:

• must pass in an existing timer
• works only for existing timers (errors out otherwise)
I believe this is regardless of what comments say - if I read 'ExtendTimer' somewhere in the code (or even use it because IDE auto-completes) there will be no reference to documentation. Having 'ExtendTimer' do a 'StartTimer' if the timer is not running is unexpected.

Could we make it work as 'timer must exist and running' to match expectations? Alternatively to have a different API of sorts (unsure how, but maybe a 'get timer' and have operations on the timer itself of 'extend')?

Ideally code should follow a path of least surprise / self-documenting.
```

I've looked at lots of different Timer APIs in the past, and to me, this one seems beyond confusing. My general feedback is:
• It's using "App state" instead of a timer object
• Why not just have a timer object? When you start one, you get a timer. That's the thing you cancel later, or can extend?

That way this would be pretty both developer familiar and friendly, but also super self-documenting. Let's fix Timers, Start and Extend at the same time?

Reply:
You are right, the current design is Timer object is hidden beneath SystemLayer, when you start one, the caller does not get a timer object.

Error StartTimer(uint32_t aMilliseconds, TimerCompleteFunct aComplete, void * aAppState);
 
Right now, the upper layer has two way to identify a timer, via combination of TimerCompleteFunct  and AppState, or a Callback function, depending on how you start the timer. The benefit of this design is echo user of system timer does not need to maintain the timer object,  it can just use the completeFunc and AppState(this pointer) to identify the timer.

I agree this design looks intuitive at first glance, should we handle this refactor in a follow-up PR?  Since we need to update all the users to track the timer it started, I would like to have a broader feedback of this change.

Reply:
We also need to look at how much more RAM if we track timer with timer object directly, We have a lot of static table in our code base. For example, we have 16 ExchangeContext by default, and add a timer object to it will consume sizeof(pointer) x 16 more bytes in BSS, same apply to all other static tables which use timer.